Refactor model loading everywhere

danbraunai opened the pull request:
## Description

_This PR is another attempt at #81 ._

- Creates RunInfo classes which load training information from wandb or local paths
- All `.from_pretrained` methods now return a single instance of the class they're attached to
- Takes @mivanit 's idea of having TMS and ResidMLP models be "LoadableModule"s which must define a from_pretrained and from_run_info.
- Moves configs related to experiments to a new `configs.py` in each experiment dir. This is for general cleanliness and avoiding circular imports.
- Adds `spd/interfaces.py` to house the `RunInfo` and `LoadableModule` base classes.

Now, if you want both the config and the model, instead of:
```
config, model, _ = ComponentModel.from_pretrained(path)
```
we have
```
run_info = SPDRunInfo.from_path(run_id)
model = ComponentModel.from_run_info(run_info)
config = run_info.config
```

## Related Issue
See [this](https://opensourcemechanistic.slack.com/archives/C08N7E5KNG7/p1753404925866329) slack message

## Motivation and Context
Model loading is broken for TMS and ResidMLP. We also want a cleaner way to get data of various sorts from completed runs.

## How Has This Been Tested?
The following test (which is not in the PR), adapted from #81 , works:
```
import torch
from spd.models.component_model import ComponentModel, SPDRunInfo

CANONICAL_RUNS: dict[str, str] = {
    "tms_5-2": "wandb:goodfire/spd/runs/u9lslp82",
    "tms_5-2-id": "wandb:goodfire/spd/runs/hm77qg0d",
    "tms_40-10": "wandb:goodfire/spd/runs/pwj1eaj2",
    "tms_40-10-id": "wandb:goodfire/spd/s2yj41ak",
    "resid_mlp1": "wandb:goodfire/spd/runs/pzauyxx8",
    "ss_mlp": "wandb:spd/runs/ioprgffh",
}

for _, model_path in CANONICAL_RUNS.items():
    device = "cuda" if torch.cuda.is_available() else "cpu"
    run_info = SPDRunInfo.from_path(model_path)
    ComponentModel.from_run_info(run_info)
    ComponentModel.from_pretrained(model_path)
```
However, I had to add a really gross deprecation handling to SPDRunInfo.from_path:
```
            config_dict = yaml.safe_load(f)
            if "figures_fns" in config_dict:
                name_idx = None
                for idx, fn in enumerate(config_dict["figures_fns"]):
                    if fn["name"] == "uv_and_identity_ci":
                        name_idx = idx
                        break
                if name_idx is not None:
                    logger.warning(
                        "uv_and_identity_ci is deprecated, but has value: {val}. Removing from config."
                    )
                    del config_dict["figures_fns"][name_idx]
```

Would be nice to have something like the above tests in the repo. This can be a separate PR IMO since we don't yet have that functionality (also OK in this PR, but this PR does seem very big already). We'd want new canonical runs which don't have the old figure function. We'll eventually want a better way to handle deprecated names for figure functions, but we'll want to wait for #78 and for a config change that we want backward compatibility for.


## Does this PR introduce a breaking change?
Yes. All code that used <model>.from_pretrained will now break as the return types are different.

mivanit reviewed (approved):
left some minor comments, will leave another comment with a suggestion on handling configs/loading/saving later

danbraunai commented on spd/interfaces.py:
```diff
@@ -0,0 +1,46 @@
+
+
+@dataclass
+class RunInfo[T]:
```
Made it an ABC and made from_path an abstractmethod.

I might be missing something but I don't think the configs do have a lot in common. They are all BaseModels, so I guess we could bound it to that. But I'm not even sure we want to make that restriction, it's probably fine if someone just wants to use a dataclass instead.

mivanit replied:
On second though, not sure if this is worth it. not that much commonality between different versions of `from_path`, I think the current state works fine

~~I think we could have `from_path` on `RunInfo` actually be implemented, and we'd only need to override it for the Resid MLP one. Could look something like:~~

```python
    def from_path(cls, path: ModelPath) -> "RunInfo[T]":
        """Load the run info from a wandb run or a local path to a checkpoint."""
        train_cfg_fname = f"{cls.name}_train_config.yaml"
        model_fname = f"{cls.name}.pth"
        if isinstance(path, str) and path.startswith(WANDB_PATH_PREFIX):
            # Check if run exists in shared filesystem first
            run_dir = check_run_exists(path)
            if run_dir:
                # Use local files from shared filesystem
                train_config_path = run_dir / train_cfg_fname
                checkpoint_path = run_dir / model_fname
            else:
                # Download from wandb
                wandb_path = path.removeprefix(WANDB_PATH_PREFIX)
                api = wandb.Api()
                run: Run = api.run(wandb_path)
                run_dir = fetch_wandb_run_dir(run.id)
                train_config_path = run_dir / train_cfg_fname
                checkpoint_path = run_dir / model_fname
        else:
            # `path` should be a local path to a checkpoint
            train_config_path = Path(path).parent / train_cfg_fname
            checkpoint_path = Path(path)

        with open(train_config_path) as f:
            train_config_dict = yaml.safe_load(f)

        train_config = T(**train_config_dict) # this is not strictly allowed but we can infer the cls
        return cls(checkpoint_path=checkpoint_path, config=train_config)
```

mivanit replied:
I also think `InductionModelPaths` and `ResidualMLPPaths` are somewhat overkill, and their purpose isn't clear to me.

danbraunai commented on spd/experiments/tms/tms_decomposition.py:
```diff
@@ -25,10 +25,10 @@ def save_target_model_info(
     save_to_wandb: bool,
     out_dir: Path,
```
Refactored in https://github.com/goodfire-ai/spd/pull/94/commits/3965af31c7c6f9a6c6b8810f8f99c5f7143ca790

danbraunai commented on spd/experiments/tms/models.py:
```diff
@@ -1,42 +1,58 @@
-    checkpoint: Path
+    @override
+    @classmethod
+    def from_path(cls, path: ModelPath) -> "TMSTargetRunInfo":
```
Refactored in https://github.com/goodfire-ai/spd/pull/94/commits/3965af31c7c6f9a6c6b8810f8f99c5f7143ca790. But note the other comment about not a 1-1 mapping between "exp_name" and names given to trained models. I think the current "task_name" should be standardized and be a our canonical thing we use here. Unfortunately it uses residual_mlp instead of resid_mlp so this would be a breaking changes. Maybe that's fine though.

mivanit replied:
agree that the current task name should be standardized -- I think breaking change is probably fine?

danbraunai replied:
Moved breaking change to #95

danbraunai commented:
@mivanit Thanks for the review, lots of good pickups there.

I've addressed/commented on all of the comments part of the review itself. As for the later comment:

> ResidualMLPConfig should definitely at least be called ResidualMLPModelConfig for consistency, even if we don't do any inheritance

Eh yeah I guess. Changed in 5c2a434. Fwiw I named it this way because I thought "ResidualMLP" is already just type of model, and doesn't presume a particular dataset. TMS is more a model + a special dataset. But yeah agree this is clearer now with so many classes.

I think I'll want to consider making breaking changes for resid_mlp before merging. In particular:
1. Renaming task_name from residual_mlp to resid_mlp. This will let me use the "task_name" as the canonical thing that gets used for target model saving. Possibly other benefits. Maybe even changing all the classnames from ResidualMLP... to ResidMLP...
2. Renaming some variables. E.g. resid_mlp_config in ResidualMLPTrainConfig to resid_mlp_model_config.

> ClassVar[ConfigDict] = ConfigDict(extra="forbid", frozen=True)
This is a special pydantic [thing](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict):
- extra="forbid" will raise an error if someone passes in an extra config arg. This is great because quite often someone will have something in their config thinking that it's being used when it's actually not. Have seen this multiple times in the past.
- frozen=True is nice because it means that you can't change any of the attributes in the config throughout the program execution. So we know that the config in the beginning is the config in the end. Though note that we do have `general_utils.replace_pydantic_config()` which allows creating a new object based on that config.

We could have these classes inherit from something that always uses this same ConfigDict, but I don't think the added complexity is worth it. I also think it's a nice reminder for people to learn about ConfigDict and what it's doing

<>TrainConfig share a lot of parameters, most notably the wandb_project which is something we should enforce any TrainConfig has

Hmm maybe. This PR might not show it, but I do think we pay a non-trivial cost to have thing inherit from (abstract) base classes. I kind of like that all of our config classes just inherit from BaseModel so people don't have to click around to find out what they are. I also don't think we're at a big risk of people not defining a wandb_project in their train config and screwing things up. If it was several very obvious arguments then I'd be more in favour. But even things like lr we wouldn't want to enforce. Nice having things in one place.

> <>TaskConfig share task_name, which should be attached to the class itself and not instances
This is the standard syntax for pydantic [discriminated unions](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-str-discriminators) which we use here. Don't think we want to change that.

> <>ModelConfig dont have a lot in common, and maybe this doesn't need to inherit from a common base class

agreed


Let me know what you think. Some disagreements and open-ended comments there. But it seems like we've mostly converged on this which is nice.

danbraunai commented:
@mivanit I decided against making the breaking changes with resid_mlp renaming and simpler saving/loading based on "task_name". As it is, we can have this PR be non-breaking and then a new PR which will do all the breaking stuff (and be easier to review).

So do let me know what you think of this PR and I'll take a final look and merge before doing the renaming PR.

mivanit commented on spd/experiments/resid_mlp/resid_mlp_decomposition.py:
```diff
@@ -77,23 +62,20 @@ def main(
+        train_config=target_run_info.config,
+        model_name="resid_mlp",
     )
+    save_file(target_run_info.label_coeffs.detach().cpu().tolist(), out_dir / "label_coeffs.json")
```
I don't love that sometimes we save things in a way that doesn't pass through `save_run_info`, but it's probably not a big deal

mivanit commented:
> I decided against making the breaking changes with resid_mlp renaming and simpler saving/loading based on "task_name". As it is, we can have this PR be non-breaking and then have the separate https://github.com/goodfire-ai/spd/pull/95 which will do all the breaking stuff (and be easier to review).

delaying breaking changes seems fine to me, I just think eventual consistency would be nice

> extra="forbid" will raise an error if someone passes in an extra config arg.

This makes sense! I would have assumed this is true by default, and it's not immediately clear why this wouldn't be a decorator on the class instead -- but this confusion on my part is due to my lack of familiarity with pydantic, haha

>  I kind of like that all of our config classes just inherit from BaseModel so people don't have to click around to find out what they are. 

I would say that even if all `<>TrainConfig`s inherit from a `TrainConfig(BaseModel)` which doesn't define a single parameter, there would still be a benefit in terms of making it clear that "all these configs belong to the same category, in a sense". Not a super important point though, and definitely doesn't need to be in this PR.

mivanit commented on spd/experiments/lm/component_viz.py:
was this deleted intentionally? feels unrelated to the main PR

danbraunai replied:
ty. I since deleted these on dev to make it clearer to people who want to go back and find it. We hadn't used them for a long time.

mivanit commented on spd/experiments/lm/plot_embedding_components.py:
also flagging for possible unintentional deletion

danbraunai replied:
See above.

mivanit reviewed (approved):
LGTM, but I flagged possible unintentional deletion of `component_viz.py` and `plot_embedding_components.py`

danbraunai commented:
> even if all <>TrainConfigs inherit from a TrainConfig(BaseModel) which doesn't define a single parameter, there would still be a benefit in terms of making it clear that "all these configs belong to the same category, in a sense"

In this case my head just says "do anything here, just make sure it's a BaseModel", which I think is fine. Agree that since we have a consistent naming scheme this implies that they're all part of the same category though. I think we might just differ a bit on our perceived cost/benefit induced by inheritance in general.
